### ISSUE TITLE
Unified naming scheme for packages

### ISSUE THREAD

**Issue by alexanderkoller** [dialogos-project/dialogos#82]
Currently, the source code is spread over a wild mixture of package names (based on CLT, Saarland U, and CMU for various historical reasons). We should switch to a unified naming scheme for packages (in dialogos and all the plugins). "app.dialogos...." seems like a reasonable choice.

**Comment by timobaumann** [comment]
I think all of dialogos itself is com.clt.* and just the plugins use different namespaces (which is not a problem).
I would argue against changing the main namespace as it will break whatever plugins others have developed based on our code. (And I know at least 1 such plugin.)

**Comment by timobaumann** [comment]
what's the advantage of 
```
app.dialogos.
```
over
```
com.clt.
```
? It's not even shorter. 

If at all, we should change it before release. However, I don't clearly see the highlights yet.

**Comment by alexanderkoller** [comment]
"com.clt" indicates a company that no longer exists. We will never make new packages that start with "com.clt", and will therefore perpetuate a messy package structure.

I'm not particularly married to "app.dialogos", but I do think there should be a unified package-name prefix.

**Comment by akoehn** [comment]
I don't really care either way but there is plenty of precedent for continuing to use a namespace after a change of ownership.  Even Oracle has packages with very weird names due to backward-compatibility.  For example, the Java Berkeley-DB package uses ```com.sleepycat```.

So I would just not change the namespace and see it as an acknowledgment that this code originates from clt.  I also see no reason why new code in the core dialogos code shouldn't also be packaged under `com.clt`.

But I won't do the work and I don't have code depending on dialogos so it is up to you two to decide :-)

**Comment by alexanderkoller** [comment]
Yes, but as you say, that was for backward compatibility. Once there is a ton of code that links against your library, you can no longer change names without breaking the Internet.

The empirical fact is that new code in the core DialogOS code was _not_ packaged under com.clt, but both Timo and I put our own universities in. This just seems arbitrary and messy.

I have a two-hour train ride this afternoon and will try to get it done. It would be good if nobody would commit changes between 16:00 and 20:00 CEST today. :)

**Comment by timobaumann** [comment]
I think plugins are different from the main application and can easily be in their own namespaces without breaking anything.

One compatibility issue is existing external clients (like the bulb-client and everything else that people have written), like the DB client we used in Schnupperstudium, and whatever other people came up with. That was a public API and I don't think we should change it. I've invested quite a bit in making old XML files work, so it might be unwise to annoy existing "power users". 

Two hours on the train could also be used for testing existing dialogs, finding inconsistencies in the UI, finding unexpected behaviour or reworking the Wiki...

If you're really in changing mood, please make it just "dialogos" rather than "app.dialogos". The domain-as-package name is so 90ies. and it's unclear how long there'll be funding for dialogos.app anyway.

Otherwise, if you're in name-changing mood, you could rename the plugin folders (no need for Dialogos_ in the name). Lots of things that are more relevant than the package name. And that don't cause trouble to users of the code.

My estimate is that I will have to spend 1 more hour changing code on our end and then probably 4 hours communicating the changes, speaking with people and trying to get their setups fixed.

**Comment by alexanderkoller** [comment]
Fair point about the plugins. I will leave the plugins outside of the core dialogos repository alone for now.

I really don't get the point about "power users". How much code that relies on the names of our plugin and client classes do you think exists in the world? And given that these were developed by programmers, what effort is involved in just changing the package name in the input? DB clients will be less important than they used to, given that we now have an official sqlite plugin; and I can of course fix our own bulb client too.

The Java package naming conventions are still the same as in the 1990s, so "app.dialogos" is technically correct, but I could live with just "dialogos" too.

I am not particularly in a "name changing" mood. I am in a "let's break things before the first release, and before there are more power users than now, rather than later" mood.

The one thing that occurs to me right now that would keep me from making the change is if the fully qualified class names are written literally in the saved DialogOS files, because this would break all existing DialogOS models. Does someone know if they are?

**Comment by timobaumann** [comment]
Qualified names are in all model files. There's code for switching over from commercial-based node names for ASR/TTS to our current node names. That would have to be extended to deal with models that have been built in the meantime and for all types of node names.

It may be just me who has downstream changes, but I have about 8 people with models and code checkouts and a lot of different model files that we use at different times. You're just causing me to do extra work and I really don't see the advantage at all. 

I knew there was going to be work between using the unofficial repository and switching to the public repo. That was anticipated. I find it unfair to introduce this now, at no advantage.

Re plugins: The question is whether users (teachers) still have their programmers (students 10 years ago) around to fix things if their beloved plugin breaks. I just don't see why we need to introduce such backward incompatibility.

**Comment by timobaumann** [comment]
There's also another customer/user (company in Denmark I'm collaborating with on the SMOOTH project) who is relying on the current names. Again: they surely can change it but they also can spend their budget in more useful ways.

**Comment by timobaumann** [comment]
There's also another customer/user (company in Denmark I'm collaborating with on the SMOOTH project) who is relying on the current names. Again: they surely can change it but they also can spend their budget in more useful ways.

**Comment by alexanderkoller** [comment]
As I said: I'd rather do the breaking change now, while there is a limited number of DialogOS users, rather than in a year or two when more people will be affected. We have a unique opportunity right now for code cleanup that will not come back.

I understand that it causes an extra effort for programmers; this seems annoying but not catastrophic.

But if the change breaks model files, that is catastrophic, because it also affects other users who have no idea how to recover from this error. For that reason, I will agree with you that we can't change package names that existed in commercial DialogOS. I will also agree that if this is controversial, and so upsetting to you, we should talk in more detail about the ramifications before we make the change.

However, I would then suggest that we at least change the package names of the official plugins to com.clt names so they are consistent with the rest of the system. Both the Sphinx and the Mary plugin currently have package names that are historical accidents (de.saar and edu.cmu, respectively). There really should not be a lot of code or models that were developed against these package names, right?

By the way: If someone has old-style DialogOS models, will they now break because the ASR and TTS are now different classes than before? Can we offer some sort of import functionality so they can be loaded by layman users?

**Comment by timobaumann** [comment]
there's glue code for loading old models, see https://github.com/dialogos-project/dialogos/issues/56.
If you go changing plugin namespaces, please adapt the glue code to account for the intermediate version and test the example files in the other plugin repositories. 

I think we've been discussing that earlier would have been better than later and I fully agree. I also agree that earlier-than-later will still incur some extra work and I accept that.

What we have not discussed is whether there's any merit to a namespace change at all. I have at least not heard compelling arguments why this would be advantageous. There are my arguments against and there is precedent that legacy-namespaces also occur in other parts of the Java ecosystem. I'd be interested in the "highlighting". What's your sales-pitch for the release notes? What value is added by a uniform namespace?

**Comment by timobaumann** [comment]
I'm taking this off v2.0 because we're starting the release process now. this can st

**Comment by alexanderkoller** [comment]
Fair enough.
